### PR TITLE
rust cli

### PR DESCRIPTION
**Pull request by burgerdev**
First impression of a concise Rust implementation of the CLI - turned out I had this sitting on an abandoned branch.

### CODE REVIEW

**Comment by pkern** [comment]
Generally looks good to me. I have nothing to add to your TODOs. Thanks! :)

File: rust/Cargo.toml [hunk 1]
```diff
@@ -7,17 +7,29 @@ edition = "2021"
...
 
 [lib]
 name = "glome"
 path = "src/lib.rs"
+
+[[bin]]
+name = "glome"
+path = "src/cli/bin.rs"
+
+required-features = ["cli"]
```
Comment (burgerdev) [inline rust/Cargo.toml:35]: Document what this does.

File: rust/src/cli/bin.rs [hunk 2]
```diff
@@ -0,0 +1,136 @@
+use base64::{engine::general_purpose, Engine as _};
+use clap::{Args, Parser, Subcommand};
+use std::convert::TryInto;
+use std::fs;
+use std::io::{self, Read, Write};
+use std::path::PathBuf;
+use x25519_dalek::{PublicKey, StaticSecret};
```
Comment (burgerdev) [inline rust/src/cli/bin.rs:7]: Can we make this crypto-lib-agnostic?

File: rust/src/cli/bin.rs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    let (_, challenge) = args.challenge.split_at(challenge_start + 3);
+    let parts: Vec<_> = challenge.split("/").collect();
+    if parts.len() != 4 || parts[3] != "" {
+        panic!("unexpected format") // TODO(burgerdev): better error output
+    }
+    let mut handshake = general_purpose::URL_SAFE.decode(parts[0]).unwrap(); // TODO(burgerdev): return the error
+    if handshake.len() < 33 {
+        panic!("handshake too short") // TODO(burgerdev): better error output
+    }
+    let _message_tag_suffix = handshake.split_off(33);
```
Comment (burgerdev) [inline rust/src/cli/bin.rs:112]: Suffix? Prefix?

**Review by burgerdev — COMMENTED** [review]
Thanks for the review! I'll leave a couple more TODOs :)

File: rust/Cargo.toml (hunk 1 above)
Comment (burgerdev) [reply rust/Cargo.toml:40]: Done

File: rust/src/cli/bin.rs (hunk 2 above)
Comment (burgerdev) [reply rust/src/cli/bin.rs:9]: Leaving this open for a future improvement.

File: rust/src/cli/bin.rs (hunk 3 above)
Comment (burgerdev) [reply rust/src/cli/bin.rs]: Renamed

File: rust/src/cli/bin.rs [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+    }
+    let mut handshake = general_purpose::URL_SAFE.decode(parts[0])?;
+    if handshake.len() < 33 {
+        return Err("handshake too short".into());
+    }
+    let message_tag_prefix = handshake.split_off(33);
+    let raw_public_key: [u8; 32] = handshake
+        .split_off(1)
+        .try_into()
+        .expect("there should be exactly 33 byte in the argument");
```
Comment (pkern) [inline rust/src/cli/bin.rs:129]: bytes

File: rust/src/cli/bin.rs [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+    let theirs: PublicKey = raw_public_key.into();
+
+    // Check public key prefix, if present.
+    let prefix = handshake[0];
+    if prefix & 1 << 7 == 0 {
+        let pubkey = ours.public_key().to_bytes();
+        if pubkey[31] != prefix {
+            return Err("own public key does not match requested prefix".into());
```
Comment (pkern) [inline rust/src/cli/bin.rs:137]: Could we include a debugging hint?

**Review by pkern — CHANGES_REQUESTED** [review]
Only one typo and a question about errors... Thank you very much, otherwise LGTM

File: rust/src/cli/bin.rs (hunk 4 above)
Comment (burgerdev) [reply rust/src/cli/bin.rs]: done

File: rust/src/cli/bin.rs (hunk 5 above)
Comment (burgerdev) [reply rust/src/cli/bin.rs]: done

**Review by pkern — APPROVED** [review]
LGTM.

It'd be good to also have a presubmit GitHub Action, but let's leave that for a followup PR.

**Pull request merged**